### ISSUE TITLE
(performance) Reduce huge query on searching for patient names (Query Rank #4)

### ISSUE THREAD

**Issue by sfount** [third-culture-software/bhima#1373]
This is a profile of the most expensive queries currently running on an installation database. 
```
# Profile                                                                                                                                   
# Rank Query ID           Response time   Calls R/Call   V/M   Item                                                                         
# ==== ================== =============== ===== ======== ===== ===========                                                                  
#    1 0x86863BB0CF7F56D2 4834.8091 34.2%   106  45.6114 22.87 CALL CalculateCashInvoiceBalances                                            
#    2 0x172C9AD568506549 2271.6316 16.1%    17 133.6254 15.85 SELECT cash voucher project debtor patient cash_box user                     
#    3 0x7146126EBA8A45BB 1906.9753 13.5%    78  24.4484 13.47 CALL PostInvoice                                                             
#    4 0x79B3A6B3AFBA6A34 1073.9583  7.6%  1059   1.0141  0.19 SELECT patient project debtor debtor_group village sector user patient_visit 
#    5 0x70FBFAE700A0D0C7  879.2873  6.2%   106   8.2952 82.17 CALL PostCash                                                                
#    6 0x9A1C8B26FB9FDA98  737.8403  5.2%    95   7.7667  0.13 SELECT UNION combined_ledger invoice project                                 
#    7 0x006CA608F6F95299  665.7194  4.7%   165   4.0347  0.28 SELECT combined_ledger document_map                                          
#    8 0x3183E320043295DC  445.9682  3.2%   116   3.8446  0.06 SELECT combined_ledger                                                       
#    9 0x6F6D9FDC88980472  229.7403  1.6%     2 114.8702  2.17 SELECT combined_ledger invoice voucher project                               
#   10 0x4E40D93B961475C4  228.1232  1.6%    79   2.8876 28.06 CALL WriteInvoice                                                            
#   11 0x04BEBBE2B4CDCF36  196.6480  1.4%    49   4.0132  0.14 SELECT combined_ledger                                                       
#   12 0xB97BDB4EDF4B0656  179.1795  1.3%    12  14.9316  0.04 SELECT debtor_group debtor combined_ledger                                   
#   13 0xAAA3A0523A8998D1  176.1755  1.2%    12  14.6813  0.05 SELECT debtor_group debtor combined_ledger account                           
# MISC 0xMISC              309.3979  2.2%  2030   0.1524   0.0 <58 ITEMS>  
```

Query number \#4 is repsonsible for **7.6%** of MySQLs CPU time over the course of the log, it is called a total of **1059** times. 

A more detailed view can be found here: 
```sql
# Query 4: 0.07 QPS, 0.07x concurrency, ID 0x79B3A6B3AFBA6A34 at byte 2190710                                                               
# This item is included in the report because it matches --limit.                                                                           
# Scores: V/M = 0.19                                                                                                                        
# Time range: 2017-03-20 07:49:42 to 11:52:03                                                                                               
# Attribute    pct   total     min     max     avg     95%  stddev  median                                                                  
# ============ === ======= ======= ======= ======= ======= ======= =======                                                                  
# Count         26    1059                                                                                                                  
# Exec time      7   1074s   640ms      4s      1s      2s   443ms   816ms                                                                  
# Lock time     39   974ms   765us     5ms   919us     2ms   294us   799us                                                                  
# Rows sent      0   7.76k       0      10    7.50    9.83    3.74    9.83                                                                  
# Rows examine  19  29.23M  28.15k  28.42k  28.27k  27.29k       0  27.29k                                                                  
# Query size    62   1.68M   1.62k   1.63k   1.62k   1.61k       0   1.61k                                                                  
# String:                                                                                                                                   
# Databases    bhima                                                                                                                        
# Hosts        localhost                                                                                                                    
# Users        bhima                                                                                                                        
# Query_time distribution                                                                                                                   
#   1us                                                                                                                                     
#  10us                                                                                                                                     
# 100us                                                                                                                                     
#   1ms                                                                                                                                     
#  10ms                                                                                                                                     
# 100ms  ################################################################                                                                   
#    1s  ###########################                                                                                                        
#  10s+                                                                                                                                     
# Tables                                                                                                                                    
#    SHOW TABLE STATUS FROM `bhima` LIKE 'patient'\G                                                                                        
#    SHOW CREATE TABLE `bhima`.`patient`\G                                                                                                  
#    SHOW TABLE STATUS FROM `bhima` LIKE 'project'\G                                                                                        
#    SHOW CREATE TABLE `bhima`.`project`\G                                                                                                  
#    SHOW TABLE STATUS FROM `bhima` LIKE 'debtor'\G                                                                                         
#    SHOW CREATE TABLE `bhima`.`debtor`\G                                                                                                   
#    SHOW TABLE STATUS FROM `bhima` LIKE 'debtor_group'\G                                                                                   
#    SHOW CREATE TABLE `bhima`.`debtor_group`\G                                                                                             
#    SHOW TABLE STATUS FROM `bhima` LIKE 'village'\G           
#    SHOW CREATE TABLE `bhima`.`village`\G                                                                                                  
#    SHOW TABLE STATUS FROM `bhima` LIKE 'sector'\G                                                                                         
#    SHOW CREATE TABLE `bhima`.`sector`\G                                                                                                   
#    SHOW TABLE STATUS FROM `bhima` LIKE 'user'\G                                                                                           
#    SHOW CREATE TABLE `bhima`.`user`\G                                                                                                     
#    SHOW TABLE STATUS FROM `bhima` LIKE 'patient_visit'\G                                                                                  
#    SHOW CREATE TABLE `bhima`.`patient_visit`\G                                                                                            
# EXPLAIN /*!50100 PARTITIONS*/                                                                                                             
SELECT BUID(q.uuid) AS uuid, q.project_id, q.reference, q.display_name, BUID(q.debtor_uuid) as debtor_uuid,                                 
      q.sex, q.dob, q.registration_date, BUID(q.debtor_group_uuid) as debtor_group_uuid, q.hospital_no,                                     
      q.last_visit, q.userName, q.originVillageName, q.originSectorName                                                                     
    FROM (                                                                                                                                  
      SELECT p.uuid, p.project_id, CONCAT_WS('.', 'PA', proj.abbr, p.reference) AS reference,                                               
        p.display_name, p.debtor_uuid AS debtor_uuid, p.sex, p.dob, p.father_name, p.mother_name, p.profession,                             
        p.employer, p.spouse, p.spouse_profession, p.spouse_employer, p.religion, p.marital_status, p.phone,                                
        p.email, p.address_1, p.address_2, p.origin_location_id, p.current_location_id,                                                     
        p.registration_date, p.title, p.notes, p.hospital_no, p.user_id, d.text, proj.abbr, dg.account_id,                                  
        dg.price_list_uuid as price_list_uuid, dg.is_convention, dg.locked, MAX(pv.start_date) AS last_visit,                               
        dg.uuid AS debtor_group_uuid, u.display_name AS userName, originVillage.name AS originVillageName,                                  
        originSector.name AS originSectorName                                                                                               
        FROM patient AS p                                                                                                                   
        JOIN project AS proj ON p.project_id = proj.id                                                                                      
        JOIN debtor AS d ON p.debtor_uuid = d.uuid                                                                                          
        JOIN debtor_group AS dg ON d.group_uuid = dg.uuid                                                                                   
        JOIN village as originVillage ON originVillage.uuid = p.origin_location_id                                                          
        JOIN sector AS originSector on originVillage.sector_uuid = originSector.uuid                                                        
        JOIN user AS u ON p.user_id = u.id                                                                                                  
        LEFT JOIN patient_visit AS pv ON pv.patient_uuid = p.uuid                                                                           
        GROUP BY p.uuid                                                                                                                     
      ) AS q                                                                                                                                
   WHERE LOWER(q.display_name) LIKE '%or%'  ORDER BY q.registration_date DESC LIMIT 10\G 
```

The query shown is responsible for looking up patients with the patient find component. It is searching for patients on display name and returning a maximum of 10 records. 

This query is not particularly slow in itself however given the frequency of the calls (every time a new letter is typed) it contributes to a huge % of the overall MySQL compute time. 

I recommend we take the following steps to reduce this overhead: 
1. Remove the **7** joins and a group by that is done for each and every search request (if modules actually need all of this information the component can search for the user on display name and then download all of their information once the patient is confirmed).
2. Encourage users to use the find patient by ID. Potentially find patient by name should be deprecated in favour of encouraging users to use Registry modules to search.

**Comment by jniles** [comment]
We don't actually use the JOIN on `debtor`, I don't think.  We have `d.text`, but no one uses that.  Additionally, the patient visit information doesn't need to be here - we can make that a separate query if necessary.

I'm not sure that forcing users to use the registry is a good idea - it means that anyone that will ever need the patient identifier will need registry access.  This will likely work against us rather than for us.  Plus it would be really inconvenient as a user experience.   We do need to insentivize users to use the identifiers.

**Comment by sfount** [comment]
@jniles these are all very useful observations - I think a specific filter for patients who are currently visiting would be useful but this information only needs to be included if you are searching for this. 

I will look into setting up a route that is specifically for searching with limited information returned.

**Comment by sfount** [comment]
This was addressed by #1383 and https://github.com/IMA-WorldHealth/bhima-2.X/pull/1418. Increase in performance can be found here https://github.com/IMA-WorldHealth/bhima-2.X/pull/1418#issuecomment-289005431.
